server/server.js

```diff
@@ -193,9 +200,69 @@ app.post('/genotet/upload', upload.single('file'), function(req, res) {
   });
 });
 
-/**
- * GET request handler.
- */
+app.use(cookieParser());
+
+app.post('/genotet/user', function(req, res) {
+  log.serverLog('POST user');
+
+  var type = req.body.type;
+  MongoClient.connect(mongoUrl, function(err, db) {
+    assert.equal(null, err, 'error occurs');
+    log.serverLog('connected to MongoDB');
+
+    var userInfo, data;
+    var authenticate = function() {
+      res.header('Access-Control-Allow-Origin', '*');
+      if (data.error) {
+        log.serverLog(data.error.type, data.error.message);
+        res.status(500).json(data.error);
+      } else {
+        var username = req.body.username;
+        user.authenticate(db, username, function(data) {
+          res.json(data);
+          db.close();
+        });
+      }
+    };
+    switch (type) {
+      case user.QueryType.SIGNUP:
+        userInfo = {
+          email: req.body.email,
+          username: req.body.username,
+          password: req.body.password,
+          confirmed: req.body.confirmed
+        };
+        user.signUp(db, userPath, userInfo, function(result) {
```

Comment:
You can use authenticate as a callback directly.
```js
user.signUp(db, userPath, userInfo, authenticate);

// edit the above
var authenticate(data) {
  ...
}
```
This also removes the problem of variable "data" being accessed from different levels of scopes.